File: server/internal/serverauth/rbac.go
package serverauth

//
// RBAC 权限模型(设计 v3b: 2026-09-04-client-login-brand-ux.md)
//
// 角色: super_admin(全量) / auditor(只读审计) / user(普通员工,无管理权限)。
// 权限: 声明式权限点常量 + 角色→权限映射(Go 内存表,不落 DB——权限集合
// 稳定、变更需发版,内表简单可测无 JOIN;未来需要运行时自定义角色时再演进)。
//
// 服务端强制: RequirePermission 中间件在所有 /api/admin/* 路由上执行,
// 前端隐藏菜单只是体验层;权限判定在服务端 403,不可绕过。
//

import (
	"net/http"

	"github.com/gin-gonic/gin"
	"github.com/picoaide/picoaide/internal/serverstore"
)

// 权限点清单。命名规范: <域>:<动作>。
const (
	PermUserRead        = "user:read"   // 用户列表
	PermUserWrite       = "user:write"  // 增改删用户、配额
	PermRoleAssign      = "role:assign" // 角色分配/提权降权
	PermDeptRead        = "dept:read"   // 部门
	PermDeptWrite       = "dept:write"
	PermAuthRead        = "auth:read"    // 认证配置(脱敏)
	PermAuthWrite       = "auth:write"   // 认证配置(含 client_secret)
	PermGatewayRead     = "gateway:read" // 网关/模型/渠道/价格
	PermGatewayWrite    = "gateway:write"
	PermUsageRead       = "usage:read"       // 用量报表
	PermQuotaWrite      = "quota:write"      // 配额/部门预算
	PermMarketRead      = "market:read"      // 市场技能
	PermMarketWrite     = "market:write"     // 技能审批/授权
	PermCapabilityRead  = "capability:read"  // 能力中心
	PermCapabilityWrite = "capability:write" // 能力中心审批
	PermConnectorRead   = "connector:read"   // 连接器
	PermConnectorWrite  = "connector:write"
	PermAuditRead       = "audit:read"            // 审计日志
	PermAuditRetention  = "audit:retention:write" // 审计保留策略(仅 super_admin)
	PermBrandRead       = "brand:read"            // 品牌
	PermBrandWrite      = "brand:write"
	PermPortalRead      = "portal:read" // 门户首页
	PermPortalWrite     = "portal:write"
	PermServerInfoRead  = "server-info:read"      // 服务器信息
	PermErrorMonRead    = "error-monitoring:read" // 错误监控
)

// AllPermissions is the full permission set (super_admin).
var AllPermissions = []string{
	PermUserRead, PermUserWrite, PermRoleAssign,
	PermDeptRead, PermDeptWrite,
	PermAuthRead, PermAuthWrite,
	PermGatewayRead, PermGatewayWrite,
	PermUsageRead, PermQuotaWrite,
	PermMarketRead, PermMarketWrite,
	PermCapabilityRead, PermCapabilityWrite,
	PermConnectorRead, PermConnectorWrite,
	PermAuditRead, PermAuditRetention,
	PermBrandRead, PermBrandWrite,
	PermPortalRead, PermPortalWrite,
	PermServerInfoRead, PermErrorMonRead,
}

// AuditorPermissions is the read-only triple allowed to the auditor role.
var AuditorPermissions = []string{
	PermAuditRead, PermUsageRead, PermUserRead,
}

// rolePerms maps role → permission set (Go in-memory table).
var rolePerms = map[string][]string{
	serverstore.RoleSuperAdmin: AllPermissions,
	serverstore.RoleAuditor:    AuditorPermissions,
	serverstore.RoleUser:       {},
}

// PermissionsOf returns the permission set for a role.
func PermissionsOf(role string) []string {
	if perms, ok := rolePerms[role]; ok {
		return perms
	}
	return nil
}

// HasPermission reports whether the user's role grants the permission.
// Pure function, no DB access.
func HasPermission(u *serverstore.User, perm string) bool {
	if u == nil {
		return false
	}
	for _, p := range PermissionsOf(u.Role) {
		if p == perm {
			return true
		}
	}
	return false
}

// RequirePermission returns a gin middleware enforcing the given permission.
// It must run after AdminAuth (which populates the admin_user context).
// Denied → 403 FORBIDDEN (never 404: audit-friendly and cannot be probed
// through menu enumeration — the 404 page is a frontend affordance only).
func RequirePermission(perm string) gin.HandlerFunc {
	return func(c *gin.Context) {
		u := currentAdmin(c)
		if u == nil {
			WriteError(c, http.StatusUnauthorized, "AUTH_REQUIRED", "未登录")
			return
		}
		if !HasPermission(u, perm) {
			WriteError(c, http.StatusForbidden, "FORBIDDEN", "没有权限执行该操作")
			return
		}
		c.Next()
	}
}

// ---------------------------------------------------------------------------
// 路由表收敛: 全部 /api/admin/* 路由必须通过 AdminRoute 注册, 显式声明
// 权限点, 杜绝「漏挂权限 → fall-open 越权」。请求序列:
//   AdminAuth(会话+CSRF) → RequirePermission(perm) → handler
//
// login 与 auth/methods 保持公开(不经过 AdminAuth); me/logout 仅需有效
// 管理会话(perm 为空串 = 登录即可)。
// ---------------------------------------------------------------------------

// adminRoutePerm records one registered admin route's required permission.
type adminRoutePerm struct {
	method string
	path   string
	perm   string
}

// adminRoutes is the in-memory registry of every AdminRoute registration;
// the integrity test compares it against gin's actual route table to prove
// no /api/admin/* route can exist without an explicit permission (fall-open
// protection). Routes registered here are also visible via Routes().
var adminRoutes []adminRoutePerm

// AdminRoute registers an admin route with an explicit required permission.
// perm == "" means "authenticated admin session only" (me/logout).
// The registry records the FULL path (group base + route) so integrity tests
// can match against gin's complete route table.
func AdminRoute(g *gin.RouterGroup, method, path, perm string, handlers ...gin.HandlerFunc) {
	chain := make([]gin.HandlerFunc, 0, len(handlers)+1)
	if perm != "" {
		chain = append(chain, RequirePermission(perm))
	}
	chain = append(chain, handlers...)
	g.Handle(method, path, chain...)
	fullPath := g.BasePath() + path
	if path == "" {
		fullPath = g.BasePath()
	}
	adminRoutes = append(adminRoutes, adminRoutePerm{method: method, path: fullPath, perm: perm})
}

// AdminRoutePerms exposes the registry (for the integrity test).
func AdminRoutePerms() []struct {
	Method string
	Path   string
	Perm   string
} {
	out := make([]struct {
		Method string
		Path   string
		Perm   string
	}, 0, len(adminRoutes))
	for _, r := range adminRoutes {
		out = append(out, struct {
			Method string
			Path   string
			Perm   string
		}{r.method, r.path, r.perm})
	}
	return out
}

File: server/webadmin/src/lib/rbac.ts
// RBAC 前端权限工具(设计 v3b): nav 与页面可见性由服务端下发的
// permissions 驱动——前端只是体验层, 服务端 RequirePermission 才是护栏。
// auditor 的只读模式在此收敛(横幅 + 隐藏写入口)。

export type Role = 'super_admin' | 'auditor' | 'user'

export interface MeUser {
  id?: number
  username?: string
  display_name?: string
  role?: Role
  permissions?: string[]
  // 0057: 密码/MFA 自助管理所需状态。
  password_must_change?: boolean
  password_changeable?: boolean
  mfa_enabled?: boolean
  source?: string
}

export function isAuditor(user: MeUser | null | undefined): boolean {
  return user?.role === 'auditor'
}

// ---------------------------------------------------------------------------
// 权限点常量(与服务端 serverauth/rbac.go 对齐;前端只作导航与入口可见性)。
// 2026-09-13(审计 R7 webadmin-branding-3)从 lib/nav.ts 迁到这里:nav 的 perms
// 声明与页面的能力判定必须用**同一份**常量,免得两处各写一份字符串而漂移。
// ---------------------------------------------------------------------------
export const PERM_USER_READ = 'user:read'
export const PERM_USER_WRITE = 'user:write'
export const PERM_DEPT_READ = 'dept:read'
export const PERM_DEPT_WRITE = 'dept:write'
export const PERM_AUTH_READ = 'auth:read'
export const PERM_GATEWAY_READ = 'gateway:read'
export const PERM_USAGE_READ = 'usage:read'
export const PERM_MARKET_READ = 'market:read'
export const PERM_CAP_READ = 'capability:read'
/** 能力中心写面(审批/上下架/归属转移等处置动作;应用中心复用同一点,见 §13)。 */
export const PERM_CAP_WRITE = 'capability:write'
export const PERM_CONNECTOR_READ = 'connector:read'
export const PERM_SERVERINFO_READ = 'server-info:read'
export const PERM_AUDIT_READ = 'audit:read'
/** 报表订阅列表(hook_url 是凭据本体;服务端**刻意**不发给 auditor,见 rbac.go)。 */
export const PERM_REPORT_READ = 'report:read'

// 当前登录管理员的模块级快照:App 在 /me 成功后写入(见 App.tsx),页面用
// hasPermission 做"体验层"判定 —— 隐藏拿不到的入口、**不请求必然 403 的接口**。
// 服务端 RequirePermission 才是护栏,这里不是安全边界。
let currentAdmin: MeUser | null = null

/** 记录当前管理员(登录/刷新后由 App 调用;登出传 null)。 */
export function setCurrentAdmin(user: MeUser | null): void {
  currentAdmin = user
}

/** 读取当前管理员快照(未登录/尚未拉到时为 null)。 */
export function currentAdminUser(): MeUser | null {
  return currentAdmin
}

/**
 * 当前用户是否持有权限点 perm。
 *
 * 缺省返回 **true**(放行):服务端未下发 permissions(旧版本/异常)或页面在
 * App 之外被直接渲染(组件测试)时"不知道权限",此时不该凭空收权 —— 与
 * nav.isNavVisible 的退回分支同口径。服务端 RequirePermission 仍是唯一护栏;
 * 只有**明确**下发了权限集且其中不含该点时才返回 false。
 */
export function hasPermission(perm: string, user: MeUser | null | undefined = currentAdmin): boolean {
  const granted = user?.permissions
  if (!Array.isArray(granted)) return true
  return granted.includes(perm)
}

export function roleLabel(role?: Role): string {
  switch (role) {
    case 'super_admin': return '超级管理员'
    case 'auditor': return '审计员'
    case 'user': return '普通员工'
    default: return '用户'
  }
}
